active/0013-configuration-migration.md

```diff
@@ -0,0 +1,162 @@
+# Data import and export using Mnesia backups
+
+## Change log
+
+* 2021-09-09: @k32 Initial draft
+* 2021-09-15: @k32 Reduce the scope. Add description of the UI
+
+## Abstract
+
+Backup and restore of database tables and configuration should be used for the cluster upgrade.
+This procedure should replace the JSON configuration dumps.
+
+## Motivation
+
+Upgrading the EMQ X cluster between major and minor releases requires redeployment of the cluster.
+User-defined ACL chains, user accounts, and other data is persisted in mnesia, and this data needs to be migrated to the new cluster.
+Currently runtime data is migrated by dumping it to a JSON file and importing it as described here:
+https://docs.emqx.io/en/broker/v4.3/advanced/data-import-and-export.html#data-import-and-export
+
+Encoding Erlang terms to JSON and decoding them back is a manual error-prone process.
+We propose to use binary dumps instead, starting from EMQ X 5.0.
+
+## Design
+
+A new `emqx_metadata` mnesia table should be added.
+Mnesia has a checkpoint feature that can be used to perform a backup.
+
+### Export
+
+The following steps will be used to export data:
+
+1. Activate a local mnesia checkpoint
+1. Backup the checkpoint to a BUP file using the standard mnesia backup callback module
+1. `emqx_metadata` table should be always added to the BUP file
+
+### Import
+
+A separate temporary BEAM process should be used during offline data migration.
+This approach solves several problems:
+
+1. The upgrade code can work with mnesia tables directly, using the regular data access layer APIs
+1. The upgrade code can perform potentially destructive operations with the data without risk of affecting the running applications
+1. The upgrade code can load beams files without affecting the running system
+
+The offline upgrade process can be started manually via a script, or automatically from the main EMQ X process when data import API is called.
+This process should load all the `beam` files that the regular EMQ X uses, but it won't start any of the applications.
+
+Steps of the offline migration process:
+
+1. Start `data_migration` application in a separate BEAM VM
+1. Run upgrade hooks
+1. If the previous step succeeds, create another backup using the standard procedure
+1. Stop the temporary BEAM VM
+1. Import the backup to the EMQ X cluster
+
+#### Upgrade hooks
+
+There should be an Erlang module called `emqx_upgrade` that collects all upgrade hooks.
+
+```erlang
+-module(emqx_upgrade).
+
+-export([upgrade_hooks/0]).
+
+upgrade_hooks() ->
+   %% Num | Intruduced in release | Function
+    [ {100, "5.1.0",                fun upgrade_lib_baz/2}
```

Comment:
if this is to be compared, maybe a tuple instead?
i.e. `"5.1.0"` -> `{5, 1, 0}`